### ISSUE TITLE
Unable to perform IAM role chaining on 5.0.0

### ISSUE THREAD

**Issue by dynamike** [99designs/aws-vault#471]
Here's the working aws config I had in 4.x.x
```
[profile account-b]
source_profile=account-a
role_arn=arn:aws:iam::bbbbbbbbbbbb:role/admin

[profile account-a]
mfa_serial=arn:aws:iam::aaaaaaaaaaaaa:mfa/dynamike
role_arn=arn:aws:iam::aaaaaaaaaaaaa:role/admin
```

I realize there are breaking changes around switching to `parent_profile` to be able to pass the `mfa_serial` between profiles. So in 0.5.0 I switched to 
```
[profile account-b]
parent_profile=account-a
role_arn=arn:aws:iam::bbbbbbbbbbbb:role/admin

[profile account-a]
mfa_serial=arn:aws:iam::aaaaaaaaaaaaa:mfa/dynamike
role_arn=arn:aws:iam::aaaaaaaaaaaaa:role/admin
```

I can access account-a just fine, but trying to access account-b I get 
```
AccessDenied: User: arn:aws:iam::aaaaaaaaaaaaa:user/dynamike is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::bbbbbbbbbbbb:role/admin
```

My setup is currently configured to do IAM role chaining from account-a into account-b. Let me know if there are other details I can give to help.

**Comment by FernandoMiguel** [comment]
please refer to the breaking changes in the changelog
https://github.com/99designs/aws-vault/releases/tag/v5.0.0

**Comment by dynamike** [comment]
yup I did that and as you can see I switched to `parent_profile` in my second example.

**Comment by rossmckelvie** [comment]
I also am observing this regression, even with the new changes in v5 for parent_profile. I would expect aws-vault to create a session with MFA for the data-admin profile that is then used to assume the test-s3 role, which can only be assumed by the admin role.

My config:

```ini
[profile master]
region=us-east-1
mfa_serial=arn:aws:iam::000000000000:mfa/ross.mckelvie

[profile data-admin]
role_arn=arn:aws:iam::111111111111:role/admin
parent_profile=master
region=us-east-1

[profile data-test-s3]
role_arn=arn:aws:iam::111111111111:role/third-party/test-s3
parent_profile=data-admin
region=us-east-1
```

Output:
```
$ aws-vault login data-test-s3
aws-vault: error: Login failed: Failed to get credentials for master: AccessDenied: User: arn:aws:iam::000000000000:user/ross.mckelvie is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::1111111111111:role/third-party/test-s3
	status code: 403, request id: 8fd98ef3-2beb-11ea-91fb-3f6ba555b35
```

**Comment by rossmckelvie** [comment]
Looks like this will be a hard one to fix, as parent_profile is essentially copying the config values from the parent. Beyond the loading of config, the chaining of profiles context is lost :( @mtibben any thoughts on best way to tackle this in v5? I was thinking if I set `source_profile=data-admin` in my final role, it would get the session from the source rather than creating a new session.

**Comment by mtibben** [comment]
Yeah I screwed this up - I didn't realise the AWS config specifies that `source_profile` will actually chain roles.

Fix is in #481, and I've released a release candidate for [v5.1.0-rc1](https://github.com/99designs/aws-vault/releases/tag/v5.0.1-rc1). Please test and let me know if this fixes your issue @rossmckelvie @dynamike

**Comment by zfasel** [comment]
@mtibben tested both 5.1.0-rc1 and a fresh build off of 99designs/aws-vault@34d73aeba1a84325c34ca70bf8de85ffc8e49471 and seems to break more than it fixes.

We leverage a setup similar to @dynamike and @rossmckelvie's use cases, with a mater account (where users live) and roles are assumed out of there for access to sub accounts. MFA is enforced on the user before they can assume any roles.

Using the following config (and swaping the variable **source_profile** for **parent_profile** for 5.x testing). All testing below is off of a clean state with no active sessions in the keychain on macos 10.15.2.

```
[profile parent]
mfa_serial = arn:aws:iam::<parentaccount>:mfa/<userid>

[profile adminrole]
source_profile = parent
role_arn = arn:aws:iam::<parentaccount>:role/Admin

[profile subaccount]
source_profile = adminrole
role_arn = arn:aws:iam::<subaccount>:role/SubAccountAdmin

[default]
region = us-east-1
```

In 4.7.1, behavior acts as follows:
```
$ aws-vault-v4 list
Profile                  Credentials              Sessions                 
=======                  ===========              ========                 
parent                 parent                 -                        
adminrole              parent                 -                        
subaccount                -                        -                        
default                  -                        -                        

$ aws-vault-v4 login subaccount
Enter token for arn:aws:iam::<parentaccount>:mfa/<userid>: 
```
and success.

After updating aws/.config for parent_profile, In 5.0.1, aws-vault prompts for MFA, but has the role chaining issue as described above.

```
$ aws-vault-v5 login adminrole
Enter token for arn:aws:iam::<parentaccount>:mfa/<userid>: 
<browser successfully launches>

$ aws-vault-v5 login subaccount
aws-vault: error: Login failed: Failed to get credentials for parent: AccessDenied: User: arn:aws:iam::<parrentaccount>:user/<userid> is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::<subaccount>:role/SubAccountAdmin
```

However in the testing of 5.1.0-rc1 and fix-role-chaining branch, things get worse with the same config as 5.0.1. 

```
$ aws-vault-v5-rc list
Profile                  Credentials              Sessions                 
=======                  ===========              ========                 
default                  -                        -                        
parent                 parent                 -                        
adminrole              -                 -                        
subaccount                -                        -                        

$ aws-vault-v5 login subaccount
aws-vault: error: Login failed: profile subaccount: credentials missing

$ aws-vault-v5 login adminrole
aws-vault: error: Login failed: profile adminrole: credentials missing

$ aws-vault-v5 exec parent env
<outputs valid envs>
```

LMK if I can help with any additional debugging.

**Comment by mtibben** [comment]
Hey @zfasel can you try using source_profile. parent_profile may not be useful for this scenario

**Comment by mtibben** [comment]
Also use the --debug flag for any commands failing

**Comment by zfasel** [comment]
@mtibben thanks for the guidance (was improperly assuming the direction was to use parent_profile still and clearly i didnt read into https://github.com/99designs/aws-vault/issues/471#issuecomment-570023080 correctly nor spend enough time with the diffs).

Using source_profile, the MFA setting on the root ("master" in my config example) does not get used for the first session token when chained, with debug throwing "MFA is not required by a chained profile, ignoring MFA" and subsequently "MFA already used in source credentials, ignoring MFA" (ah the finger pointing). 

When manually specifying *mfa_serial* repeatedly in each profile, role chaining works as expected. Huzzah!

Subset from --debug resulting in an access denied (due to lack of required MFA) using the example config listed in https://github.com/99designs/aws-vault/issues/471#issuecomment-570237064

```
$ aws-vault-v5-rc login subaccount
2020/01/02 11:10:30 [keyring] Considering backends: [keychain pass file]
2020/01/02 11:10:30 Loading config file /Users/zfasel/.aws/config
2020/01/02 11:10:30 Parsing config file /Users/zfasel/.aws/config
2020/01/02 11:10:30 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/01/02 11:10:30 [keyring] Found 1 results
2020/01/02 11:10:30 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/01/02 11:10:30 [keyring] Found 1 results
2020/01/02 11:10:30 [keyring] Querying keychain for service="aws-vault", keychain="aws-vault.keychain"
2020/01/02 11:10:30 [keyring] Found 1 results
2020/01/02 11:10:30 profile parent: using stored credentials
2020/01/02 11:10:30 MFA is not required by a chained profile, ignoring MFA
2020/01/02 11:10:30 profile parent: using GetSessionToken
2020/01/02 11:10:30 MFA already used in source credentials, ignoring MFA
2020/01/02 11:10:30 profile adminrole: using AssumeRole
2020/01/02 11:10:30 MFA already used in source credentials, ignoring MFA
2020/01/02 11:10:30 profile subaccount: using AssumeRole
```

**Comment by dynamike** [comment]
@zfasel Can you post your final redacted .aws/config?

**Comment by zfasel** [comment]
@dynamike a working config for my use case on 5.1.0-rc1 is as follows

```
[profile primary]
mfa_serial = arn:aws:iam::<root account>:mfa/<user>

[profile admin]
source_profile = primary
role_arn = arn:aws:iam::<root account>:role/Admin
mfa_serial = arn:aws:iam::<root account>:mfa/<user>

[profile subaccount]
source_profile = admin
role_arn = arn:aws:iam::<sub account>:role/DevUser
mfa_serial = arn:aws:iam::<root account>:mfa/<user>

[default]
region = us-east-1
```

**Comment by FernandoMiguel** [comment]
That's the v4 version parent_profile

**Comment by zfasel** [comment]
> That's the v4 version parent_profile

@FernandoMiguel Yes, per @mtibben in https://github.com/99designs/aws-vault/issues/471#issuecomment-570253219 with regards to 5.1.0-rc1 and role chaining.

**Comment by dynamike** [comment]
My situation is slight different than @zfasel. I had one less profile definition in my configuration (which was essentially @zfasel's primary and admin profile merged into one. 
```
[profile admin]                                                                                                                                                     
mfa_serial=arn:aws:iam::<root account>:mfa/<user>
role_arn=arn:aws:iam::<root account>:role/admin

[profile subaccount]
source_profile=admin                                                                                                                                                
mfa_serial=arn:aws:iam::<root account>:mfa/dynamike
role_arn=arn:aws:iam::<sub account>:role/admin
```

The above results in the following error when running things after clearing all cached sessions running v5.1.0-rc1
```
aws-vault --debug exec subaccount -- aws sts get-caller-identity 
2020/01/02 13:02:35 [keyring] Considering backends: [keychain pass file]
2020/01/02 13:02:35 Loading config file /Users/dynamike/.aws/config
2020/01/02 13:02:35 Parsing config file /Users/dynamike/.aws/config
2020/01/02 13:02:35 [keyring] Querying keychain for service="aws-vault", keychain="login.keychain"
2020/01/02 13:02:35 [keyring] Found 2 results
2020/01/02 13:02:35 [keyring] Querying keychain for service="aws-vault", keychain="login.keychain"
2020/01/02 13:02:35 [keyring] Found 2 results
2020/01/02 13:02:35 profile admin: using stored credentials
2020/01/02 13:02:35 MFA already used in source credentials, ignoring MFA
2020/01/02 13:02:35 profile admin: using AssumeRole
2020/01/02 13:02:35 MFA already used in source credentials, ignoring MFA
2020/01/02 13:02:35 profile subaccount: using AssumeRole
2020/01/02 13:02:35 Looking up keyring for 'admin'
2020/01/02 13:02:35 [keyring] Querying keychain for service="aws-vault", account="admin", keychain="login.keychain"
2020/01/02 13:02:35 [keyring] Found item "aws-vault (admin)"
aws-vault: error: exec: Failed to get credentials for subaccount: AccessDenied: User: arn:aws:iam::<root account>:user/dynamike is not authorized to perform: sts:AssumeRole on resource: arn:aws:iam::<root account>:role/admin
```

**Comment by zfasel** [comment]
@dynamike can confirm the same issue. Are you getting it as well when trying to exec as profile admin without any role chaining? 

Leveraging a similar .aws/config, `MFA already used in source credentials, ignoring MFA` comes up for a similar admin profile configuration and ignores mfa_serial.

**Comment by dynamike** [comment]
yup, I'm getting the same issue trying to use the admin profile.

**Comment by zfasel** [comment]
not going to open a new issue for it just yet as it plays into this same block of necessary fixes, but it also seems in this refactoring that SessionTokens are not being stored for profiles with both assumed roles with mfas. h/t to @dynamike for causing me to test outside of my config.

in 4.7.1, a config like @dynamike's stores an aws-vault seession in keychain for the profile upon successful MFA for future use.

after forcing 5.1.0-rc1 to accept the MFA (bypasing the `useChainedMfa` block in vault.go@186) and getting successful login, no session keychain entry is created, and MFA is consistently re-prompted.

**Comment by mtibben** [comment]
> Using the following config (and swaping the variable source_profile for parent_profile for 5.x testing).

To clear up any confusion, the `parent_profile` templates the specified profile. You still need to use `source_profile` to specify the credential source, either in the profile or in the parent. So it's not a replacement, it's in addition

**Comment by mtibben** [comment]
> Using source_profile, the MFA setting on the root ("master" in my config example) does not get used for the first session token when chained
> When manually specifying mfa_serial repeatedly in each profile, role chaining works as expected

Yes that's the expected behaviour. If you wanted to only specify mfa_serial only once in your config, you could also set `parent_profile` to specify the profile to import the mfa from

**Comment by mtibben** [comment]
@zfasel and @dynamike can you try again with https://github.com/99designs/aws-vault/releases/tag/v5.1.0-rc2.

My config below is working

```ini
[profile root]
mfa_serial=arn:aws:iam::<root>:mfa/<me>

[profile role1]
source_profile=root
role_arn=arn:aws:iam::<subaccount>:role/RoleOne
mfa_serial=arn:aws:iam::<root>:mfa/<me>

[profile role2]
source_profile=role1
role_arn=arn:aws:iam::<subaccount>:role/RoleTwo
mfa_serial=arn:aws:iam::<root>:mfa/<me>
```

**Comment by mtibben** [comment]
Whoops didn't mean to close this until I have confirmation from @zfasel and @dynamike that the issue has been resolved

**Comment by mtibben** [comment]
Assuming this has been resolved
